vllm/v1/structured_output/backend_xgrammar.py

```diff
@@ -87,11 +88,22 @@ def __init__(self, vllm_config: VllmConfig):
             )
         self.compiler = xgr.GrammarCompiler(
             tokenizer_info,
-            max_threads=8,
+            max_threads=self._get_grammar_compiler_threads(),
             cache_enabled=True,
             cache_limit_bytes=vllm.envs.VLLM_XGRAMMAR_CACHE_MB * 1024 * 1024,
         )
 
+    def _get_grammar_compiler_threads(self) -> int:
+        """Get the number of threads to use for the grammar compiler. The number of physical
+        threads is used to determine the thread count, and it is capped at 32 because having too
+        many threads does not yield additional benefits.
+
+        Returns:
+            int: The number of threads to use for the grammar compiler.
+        """
+        num_cpus = os.cpu_count()
+        return min(num_cpus / 2, 32)
```

Comment:
Why are 32 threads needed for this? It seems like overkill and poses problems when running multiple instances on one host. 

This also won't work well for systems with small numbers of cores, as we have several threads being used just for multi-process input, scheduler, model workers. I think we need to have a fixed floor where you subtract some number of "reserved vllm threads" to get an "available threads" estimate

Reply:
if it's an 8 CPU core system, this will result in 4.

fair point about multiple instances on a single host.  See my comment on the PR with a test script that helped find the sweet spot . I don't think we should hard code 32 since there's nothing magic about that.

Reply:
@Ubospica why did you move away from using `psutil` and getting the number of physical cores? did you get an error?

Reply:
basically I think we should increase the number of threads for structured output if we have an embarrassing number of cores, rather than taking up half by default

Reply:
> if it's an 8 CPU core system, this will result in 4.
> 
> fair point about multiple instances on a single host. See my comment on the PR with a test script that helped find the sweet spot . I don't think we should hard code 32 since there's nothing magic about that.

32 threads reflect the effective threads needed for typical schema sizes. This might be appropriate since we’re only setting an upper limit.

To allow multiple instances, I think we can add a global setting to limit the number of threads, just like we did in V1. But just not sure if we need to do it in this PR.